### PR TITLE
[AMDGPU] Fix GCNUpwardRPTracker. (WIP)

### PR DESCRIPTION
**Pull request by vpykhtin**
This is an early-preview, this patch is going to fix:

1. Maximum register pressure calculation at the instruction level. 
Previously max RP included both def and use of registers of an instruction. Now maximum RP includes _uses_ and _early-clobber defs_.

2. Uses were incorrectly tracked and this resulted in a mismatch of live-in set reported by LiveIntervals and tracked live reg set when the beginning of the block is reached.

Interface has changed, moveMaxPressure becomes deprecated and  getMaxPressure, resetMaxPressure functions are added. reset function seem now more consistent.

GCNUpwardRPTracker interface may seem a bit strange but I would like to clarify what I had in mind when created it.

The idea was to have a tracker that can work on non-implemented tentative schedules, that is:
1. Instruction order is maintained externally, like in array.
2. LiveIntervals and dead/kill flags aren't updated on a tentative schedule.

For this reason it has no functions to advance to next instruction, it should not use dead/kill flags. It only uses LiveIntervals information to track usage of a register with subregs and to calculate starting live register set. The idea is that _lane bitmask of a reg's use_ should be the same for any schedule.

### CODE REVIEW

**Comment by jayfoad** [comment]
> GCNUpwardRPTracker interface may seem a bit strange but I would like to clarify what I had in mind when created it.
> 
> The idea was to have a tracker that can work on non-implemented tentative schedules, that is:
> 
> 1. Instruction order is maintained externally, like in array.
> 2. LiveIntervals and dead/kill flags aren't updated on a tentative schedule.
> 
> For this reason it has no functions to advance to next instruction, it should not use dead/kill flags. It only uses LiveIntervals information to track usage of a register with subregs and to calculate starting live register set. The idea is that _lane bitmask of a reg's use_ should be the same for any schedule.

Thanks. That is useful information.

I can understand how the upward tracker can work like that, but how does the downward tracker work? When it sees a use operand, it does not know if it is the last use of that register.

**Comment by vpykhtin** [comment]
> I can understand how the upward tracker can work like that, but how does the downward tracker work? When it sees a use operand, it does not know if it is the last use of that register.

Downward tracker is different, it wasn't supposed to be used like that, and therefore it has different interface.

**Comment by piotrAMD** [comment]
Thanks - I can confirm the patches fixes two issues I observed recently, where the pressure was over-reported.

File: llvm/lib/Target/AMDGPU/GCNRegPressure.cpp [hunk 1]
```diff
@@ -166,66 +166,62 @@ static LaneBitmask getDefRegMask(const MachineOperand &MO,
...
     if (!MO.isUse() || !MO.readsReg())
       continue;
 
-    auto const UsedMask = getUsedRegMask(MO, MRI, LIS);
+    Register Reg = MO.getReg();
+    auto I = llvm::find_if(RegMaskPairs, [Reg](const RegisterMaskPair &RM) {
+      return RM.RegUnit == Reg;
+    });
+    if (I != RegMaskPairs.end())
+      continue;
```
Comment (arsenm) [inline llvm/lib/Target/AMDGPU/GCNRegPressure.cpp:185]: is_contained?

File: llvm/lib/Target/AMDGPU/GCNRegPressure.cpp [hunk 2]
```diff
@@ -166,66 +166,62 @@ static LaneBitmask getDefRegMask(const MachineOperand &MO,
...
+                                  const MachineRegisterInfo &MRI) {
   LaneBitmask LiveMask;
-  const auto &LI = LIS.getInterval(Reg);
   if (LI.hasSubRanges()) {
     for (const auto &S : LI.subranges())
       if (S.liveAt(SI)) {
         LiveMask |= S.LaneMask;
-        assert(LiveMask < MRI.getMaxLaneMaskForVReg(Reg) ||
-               LiveMask == MRI.getMaxLaneMaskForVReg(Reg));
+        assert(LiveMask < MRI.getMaxLaneMaskForVReg(LI.reg()) ||
```
Comment (arsenm) [inline llvm/lib/Target/AMDGPU/GCNRegPressure.cpp:220]: You shouldn't really be relying on the < value; you should only be making a subset equal bit mask test

File: llvm/test/CodeGen/AMDGPU/regpressure_printer.mir [hunk 3]
```diff
@@ -459,4 +433,55 @@ body:             |
...
+    ; RPU-NEXT: 0     0
+    ; RPU-NEXT: Live-out:
+    ;
+    ; RPD-LABEL: name: test_not_early_clobber_trivial
+    ; RPD: Live-in:
+    ; RPD-NEXT: SGPR  VGPR
+    ; RPD-NEXT: 0     0
+    ; RPD-NEXT: 0     1      %0:vgpr_32 = V_MOV_B32_e32 42, implicit $exec
+    ; RPD-NEXT: 0     1
+    ; RPD-NEXT: 0     2      %1:vgpr_32 = V_MOV_B32_e32 %0:vgpr_32, implicit $exec
```
Comment (jayfoad) [inline llvm/test/CodeGen/AMDGPU/regpressure_printer.mir:479]: So the downward tracker also needs to be fixed?

File: llvm/lib/Target/AMDGPU/GCNRegPressure.cpp [hunk 4]
```diff
@@ -261,15 +257,11 @@ void GCNRPTracker::reset(const MachineInstr &MI,
...
-void GCNUpwardRPTracker::reset(const MachineInstr &MI,
-                               const LiveRegSet *LiveRegsCopy) {
-  GCNRPTracker::reset(MI, LiveRegsCopy, true);
-}
-
 void GCNUpwardRPTracker::reset(const MachineRegisterInfo &MRI_,
                                const LiveRegSet &LiveRegs_) {
   MRI = &MRI_;
   LiveRegs = LiveRegs_;
+  LastTrackedMI = nullptr; // TODO: LastTrackedMI isnt' used, remove?
```
Comment (jayfoad) [inline llvm/lib/Target/AMDGPU/GCNRegPressure.cpp:264]: Just remove it

File: llvm/lib/Target/AMDGPU/GCNRegPressure.h [hunk 5]
```diff
@@ -149,24 +149,46 @@ class GCNUpwardRPTracker : public GCNRPTracker {
...
+  // move to the state just before the MI (in program order).
   void recede(const MachineInstr &MI);
 
   // checks whether the tracker's state after receding MI corresponds
-  // to reported by LIS
+  // to reported by LIS.
   bool isValid() const;
+
+  // deprecated.
+  decltype(MaxPressure) moveMaxPressure() = delete;
```
Comment (jayfoad) [inline llvm/lib/Target/AMDGPU/GCNRegPressure.h:179]: Might be simpler to move the base class moveMaxPressure into GCNDownwardRPTracker?

File: llvm/test/CodeGen/AMDGPU/regpressure_printer.mir (hunk 3 above)
Comment (vpykhtin) [reply llvm/test/CodeGen/AMDGPU/regpressure_printer.mir:479]: Yes, I'm going to fix it in a separate PR, we should have a perfect match of RP in both ways.

File: llvm/lib/Target/AMDGPU/GCNRegPressure.cpp (hunk 1 above)
Comment (vpykhtin) [reply llvm/lib/Target/AMDGPU/GCNRegPressure.cpp]: I've replaced it with _any_of_ because _is_contained_ doesn't accept predicate.

File: llvm/lib/Target/AMDGPU/GCNRegPressure.cpp (hunk 2 above)
Comment (vpykhtin) [reply llvm/lib/Target/AMDGPU/GCNRegPressure.cpp]: done.

File: llvm/lib/Target/AMDGPU/GCNRegPressure.h (hunk 5 above)
Comment (vpykhtin) [reply llvm/lib/Target/AMDGPU/GCNRegPressure.h]: done.

File: llvm/lib/Target/AMDGPU/GCNRegPressure.cpp (hunk 4 above)
Comment (vpykhtin) [reply llvm/lib/Target/AMDGPU/GCNRegPressure.cpp]: I haven't noticed it's used with the base class accessor, so I decided to leave it and delete TODO comment.

**Comment by vpykhtin** [comment]
I assume this can be submitted if there are no objections, and I'm going to fix downward tracker next.

File: llvm/lib/Target/AMDGPU/GCNRegPressure.h [hunk 6]
```diff
@@ -128,14 +128,7 @@ class GCNRPTracker {
...
 
-  GCNRegPressure getPressure() const { return CurPressure; }
-
-  // returns MaxPressure, resetting it
-  decltype(MaxPressure) moveMaxPressure() {
-    auto Res = MaxPressure;
-    MaxPressure.clear();
-    return Res;
-  }
+  const GCNRegPressure &getPressure() const { return CurPressure; }
```
Comment (jayfoad) [inline llvm/lib/Target/AMDGPU/GCNRegPressure.h:131]: Was this change intentional? Returning a ref seems less safe, because the caller has to be careful not to use it after mutating the tracker in any way.

File: llvm/lib/Target/AMDGPU/GCNRegPressure.h [hunk 7]
```diff
@@ -180,6 +190,13 @@ class GCNDownwardRPTracker : public GCNRPTracker {
 
   MachineBasicBlock::const_iterator getNext() const { return NextMI; }
 
+  // Return MaxPressure and clear it.
+  decltype(MaxPressure) moveMaxPressure() {
```
Comment (jayfoad) [inline llvm/lib/Target/AMDGPU/GCNRegPressure.h:194]: Why use decltype? It's longer than GCNRegPressure.

**Review by jayfoad — APPROVED** [review]
Looks OK to me. I have not understood every detail but you know this code much better than I do.

**Comment by vpykhtin** [comment]
> Looks OK to me. I have not understood every detail but you know this code much better than I do.

Thank you Jay, let's review those details, I find both points above valid and will fix them.

File: llvm/lib/Target/AMDGPU/GCNRegPressure.h (hunk 7 above)
Comment (vpykhtin) [reply llvm/lib/Target/AMDGPU/GCNRegPressure.h]: Done.

File: llvm/lib/Target/AMDGPU/GCNRegPressure.h (hunk 6 above)
Comment (vpykhtin) [reply llvm/lib/Target/AMDGPU/GCNRegPressure.h]: Done.
